attestation-evidence-broker/docs/design/design.md

```diff
@@ -0,0 +1,122 @@
+
+### Hypercall approach
+
+SEV(-ES) guest exits to VMM using `hypercall` and then interacts with PSP firmware to query attestation evidence, including attestation report, PEK, CEK, OCA certificate. In general, it includes the following steps:
+
+1. The guest application requests a shared memory page, initiates a hypercall, and switches from VMX non-root mode to VMX root mode.
+2. In VMX root mode, KVM sends the `GET_ID, PDH_CERT_EXPORT, ATTESTATION` command requests to PSP firmware.
+3. The shared memory page is filled with the data returned by the PSP firmware.
+4. The guest application can obtain attestation evidence by reading the data in the shared memory.
+
+Although this method can meet our requirements, it requires a lot of modifications to the guest kernel and KVM, and it is hard to be accepted by upstream.
+
+### Vsock approach
+
+In the current implementation, QEMU provides the QMP interface "query-sev-attestation-report" to query the attestation report in the host. The report is signed by PEK and contains the SHA-256 digest of the guest memory and VMSA encrypted with the LAUNCH_UPDATE. However, QEMU is not the only VMM. In order to support various VMM in different scenarios, it is necessary to design a general design to query attestation evidence from the host.
+
+The workflow of AEB is as followed:
+
+![architecture.png](architecture.png)
+
+1. The user-level application in the guest sends a request (including guest firmware handle) to AEB through Vsock.
+2. AEB requests to query attestation report, PEK, CEK, OCA certificate by calling multiple PSP firmware APIs (refer to the table above for specific API commands) and assembles these information into attestation evidence.
+3. AEB returns attestation evidence to the application in the guest.
+
+To query the attestation report in host with AEB, you need to apply the following patches, which will be sent to the community for review afterward:
+
+1. It is necessary to add a `SEV_GET_REPORT` interface in ccp driver so that AEB can execute `ioctl` on the `/dev/sev` device and then call the `SEV_GET_REPORT` interface to send the `ATTESTATION` command to the PSP firmware to query attestation report.
+
+2. In order to obtain the guest handle required by the `ATTESTATION` command sent by AEB to the PSP firmware, a new hypercall needs to be added to the KVM. The application in the Guest obtains the guest handle through hypercall, and then sends it to the AEB service through vsock. The AEB service uses the guest handle as the input parameter of the `ATTESTATION` command to interact with the PSP firmware.
+
+Note that hypercall is not the only way to obtain the guest handle, the qmp interface `query-sev` can query the guest handle as well. However, the implementation based on qemu cannot support multiple types of vmm, we will support obtaining the guest handle based on qemu in the future.
+
+# The Communication protocol between guest application and AEB
```

Comment:
s/The Communication protocol between guest application and AEB/Communication protocol/

Reply:
Add one line:
Below is the communication protocol between guest application and AEB.